wip-trampoline.md

```diff
@@ -0,0 +1,138 @@
+# Trampoline Onion Routing
+
+This document describes the additions we make to the existing onion routing
+to enable the use of trampoline hops.
+
+## Packet Structure
+
+### Trampoline Onion
+
+We define a fixed-size trampoline onion packet that can be nested inside onion
+packets. We define `N` to be the maximum number of trampoline nodes (excluding
+the sender node and the destination node).
+
+The trampoline packet always contains `2*N*FRAME_SIZE + 66` bytes.
+
+1. type: `trampoline_onion_packet`
+2. data:
+   * [`1`:`version`]
+   * [`33`:`public_key`]
+   * [`2*N*FRAME_SIZE`:`trampoline_hops_data`]
+   * [`32`:`hmac`]
+
+Note: we use the same onion format as the standard onion which wastes frames,
+but that theoretically allows another level of nesting (mind blown).
+However not really usable given the space constraints...
+
+TODO: maybe this should be version 1 of the overall onion version namespace?
+
+TODO: maybe we can do something more clever with TLV? At the cost of a second
+onion parser (trampoline onion parser)? Use the top-level version in the parser
+to discriminate?
```

Comment:
If we stick with 65 byte frames we can reuse the onion processing from the outer onion.

Reply:
Same, this was an old commit. I plan on removing `FRAME_SIZE` entirely, so it's going to be a lot nicer. I'll also squash all the commits together to hide the old commits.